go/vt/vtctl/grpcvtctldserver/server.go

```diff
@@ -2120,6 +2123,261 @@ func (s *VtctldServer) UpdateCellsAlias(ctx context.Context, req *vtctldatapb.Up
+func (s *VtctldServer) ValidateSchemaShard(ctx context.Context, req *vtctldatapb.ValidateSchemaShardRequest) (*vtctldatapb.ValidateSchemaShardResponse, error) {
+	si, err := s.ts.GetShard(ctx, req.Keyspace, req.Shard)
+	if err != nil {
+		return nil, vterrors.Wrapf(err, "GetShard(%v, %v) failed: %v", req.Keyspace, req.Shard, err)
+	}
+
+	if !si.HasPrimary() {
+		return nil, vterrors.Errorf(vtrpc.Code_FAILED_PRECONDITION, "no primary in shard %v/%v", req.Keyspace, req.Shard)
+	}
+	log.Infof("Gathering schema for primary %v", topoproto.TabletAliasString(si.Shard.PrimaryAlias))
+
+	getTabletReq := &vtctldatapb.GetTabletRequest{
+		TabletAlias: si.Shard.PrimaryAlias,
+	}
+
+	getTabletResponse, err := s.GetTablet(ctx, getTabletReq)
+	if err != nil {
+		return nil, vterrors.Wrapf(err, "GetTablet(%v) failed: %v", topoproto.TabletAliasString(getTabletReq.TabletAlias), err)
+	}
+
+	masterSchema, err := s.tmc.GetSchema(ctx, getTabletResponse.Tablet, nil, req.ExcludeTables, req.IncludeViews)
+	if err != nil {
+		return nil, vterrors.Wrapf(err, "GetSchema(%v, nil, %v, %v) failed: %v", si.Shard.PrimaryAlias, req.ExcludeTables, req.IncludeViews, err)
+	}
+
+	if req.IncludeVSchema {
+		validateVSchemaReq := &vtctldatapb.ValidateVSchemaRequest{
+			Keyspace:      req.Keyspace,
+			Shards:        []string{req.Shard},
+			ExcludeTables: req.ExcludeTables,
+			IncludeViews:  req.IncludeViews,
+		}
+
+		if _, err := s.ValidateVSchema(ctx, validateVSchemaReq); err != nil {
+			return nil, vterrors.Wrapf(err, "ValidateVSchema(%v) failed: %v", validateVSchemaReq, err)
+		}
+	}
+
+	// read all the aliases in the shard, that is all tablets that are
+	// replicating from the master
+	aliases, err := s.ts.FindAllTabletAliasesInShard(ctx, req.Keyspace, req.Shard)
+	if err != nil {
+		return nil, vterrors.Wrapf(err, "FindAllTabletAliasesInShard(%v, %v) failed: %v", req.Keyspace, req.Shard, err)
+	}
+
+	// then diff with all replicas
+	var (
+		rec concurrency.AllErrorRecorder
+		wg  sync.WaitGroup
+	)
+
+	for _, alias := range aliases {
+		if topoproto.TabletAliasEqual(alias, si.Shard.PrimaryAlias) {
+			continue
+		}
+
+		wg.Add(1)
+		go func(alias *topodatapb.TabletAlias) {
+			defer wg.Done()
+			if err := s.diffSchema(ctx, masterSchema, si.Shard.PrimaryAlias, alias, req.ExcludeTables, req.IncludeViews); err != nil {
+				rec.RecordError(err)
+			}
+		}(alias)
+	}
+	wg.Wait()
+	if rec.HasErrors() {
+		return nil, vterrors.Wrapf(rec.Error(), "schema diffs: %v", rec.Error().Error())
+	}
+	return &vtctldatapb.ValidateSchemaShardResponse{}, nil
+}
+
+// ValidateVSchema is part of the vtctlservicepb.VtctldServer interface
+func (s *VtctldServer) ValidateVSchema(ctx context.Context, req *vtctldatapb.ValidateVSchemaRequest) (*vtctldatapb.ValidateVSchemaResponse, error) {
+	vschm, err := s.ts.GetVSchema(ctx, req.Keyspace)
+	if err != nil {
+		return nil, vterrors.Wrapf(err, "GetVSchema(%s) failed: %v", req.Keyspace, err)
+	}
+	var (
+		shardFailures concurrency.AllErrorRecorder
```

Comment:
same comment about matching style, i'm going to pause reviewing here, as i am going to guess a lot of my commentary will continue in similar fashion for the rest of the PR